### PR TITLE
vndr github.com/docker/docker from components/engine

### PR DESCRIPTION
**Pull request by andrewhsu**
Found differences between the engine and the cli vendor of engine in the docker-ce 17.06 branch. This PR will vendor github.com/docker/docker from `components/engine` dir.

### CODE REVIEW

File: components/cli/vendor/github.com/docker/docker/opts/config.go [hunk 1]
```diff
@@ -1,98 +0,0 @@
-package opts
```
Comment (thaJeztah) [inline components/cli/vendor/github.com/docker/docker/opts/config.go]: this was removed in https://github.com/moby/moby/pull/33296

File: components/cli/vendor/github.com/docker/docker/opts/opts.go [hunk 2]
```diff
@@ -2,13 +2,11 @@ package opts
 
 import (
 	"fmt"
-	"math/big"
```
Comment (thaJeztah) [inline components/cli/vendor/github.com/docker/docker/opts/opts.go]: https://github.com/moby/moby/pull/33296

File: components/cli/vendor/github.com/docker/docker/opts/throttledevice.go [hunk 3]
```diff
@@ -1,111 +0,0 @@
-package opts
```
Comment (thaJeztah) [inline components/cli/vendor/github.com/docker/docker/opts/throttledevice.go]: https://github.com/moby/moby/pull/33296

File: components/cli/vendor/github.com/docker/docker/pkg/pools/pools.go [hunk 4]
```diff
@@ -17,15 +17,16 @@ import (
 	"github.com/docker/docker/pkg/ioutils"
 )
 
+const buffer32K = 32 * 1024
```
Comment (thaJeztah) [inline components/cli/vendor/github.com/docker/docker/pkg/pools/pools.go]: https://github.com/docker/docker-ce/pull/76

File: components/cli/vendor/github.com/docker/docker/pkg/gitutils/gitutils.go [hunk 5]
```diff
@@ -15,18 +15,24 @@ import (
 	"github.com/pkg/errors"
 )
 
+type gitRepo struct {
```
Comment (thaJeztah) [inline components/cli/vendor/github.com/docker/docker/pkg/gitutils/gitutils.go]: https://github.com/docker/docker-ce/pull/100

File: components/cli/vendor/github.com/docker/docker/pkg/term/tc_solaris_cgo.go [hunk 6]
```diff
@@ -13,7 +13,7 @@ import (
 import "C"
 
 // Termios is the Unix API for terminal I/O.
-// It is passthrough for syscall.Termios in order to make it portable with
```
Comment (thaJeztah) [inline components/cli/vendor/github.com/docker/docker/pkg/term/tc_solaris_cgo.go]: https://github.com/moby/moby/pull/33269

File: components/cli/vendor/github.com/docker/docker/registry/config_unix.go [hunk 7]
```diff
@@ -21,5 +21,5 @@ func cleanPath(s string) string {
 
 // installCliPlatformFlags handles any platform specific flags for the service.
 func (options *ServiceOptions) installCliPlatformFlags(flags *pflag.FlagSet) {
-	flags.BoolVar(&options.V2Only, "disable-legacy-registry", false, "Disable contacting legacy registries")
+	flags.BoolVar(&options.V2Only, "disable-legacy-registry", true, "Disable contacting legacy registries")
```
Comment (thaJeztah) [inline components/cli/vendor/github.com/docker/docker/registry/config_unix.go]: https://github.com/moby/moby/pull/33629 (cherry-pick: https://github.com/docker/docker-ce/pull/71)

File: components/cli/vendor/github.com/docker/docker/volume/volume.go [hunk 8]
```diff
@@ -146,7 +146,9 @@ func (m *MountPoint) Cleanup() error {
 
 // Setup sets up a mount point by either mounting the volume if it is
 // configured, or creating the source directory if supplied.
-func (m *MountPoint) Setup(mountLabel string, rootUID, rootGID int) (path string, err error) {
+// The, optional, checkFun parameter allows doing additional checking
+// before creating the source directory on the host.
+func (m *MountPoint) Setup(mountLabel string, rootUID, rootGID int, checkFun func(m *MountPoint) error) (path string, err error) {
```
Comment (thaJeztah) [inline components/cli/vendor/github.com/docker/docker/volume/volume.go]: https://github.com/moby/moby/pull/33330

**Comment by thaJeztah** [comment]
I linked the related PR's either upstream (moby/moby) and/or the cherry-picks in components/engine

File: components/cli/vendor/github.com/docker/docker/pkg/gitutils/gitutils.go (hunk 5 above)
Comment (andrewhsu) [inline components/cli/vendor/github.com/docker/docker/pkg/gitutils/gitutils.go]: this change we definitely need to get in to fix stuff to do stuff

**Comment by andrewhsu** [comment]
I'm ok to have this PR in as a stop gap measure for now to get these guys in sync, but we have to think about a way to get the vendor dir updated on a more regular basis.

**Comment by andrewhsu** [comment]
@tiborvass fyi

**Comment by thaJeztah** [comment]
By now we probably should do the same again (various PR's were merged)?

**Comment by tiborvass** [comment]
@vieux @andrewhsu @thaJeztah wdyt of https://github.com/docker/docker-ce/compare/17.06...tiborvass:fix-vendor ?

**Comment by tiborvass** [comment]
I opened my version at https://github.com/docker/docker-ce/pull/148

File: components/cli/vendor/github.com/docker/docker/pkg/pools/pools.go (hunk 4 above)
Comment (andrewhsu) [inline components/cli/vendor/github.com/docker/docker/pkg/pools/pools.go]: this change to `components/cli` was a result of a missing `vendor` update when #76 was cherry-picked to `components/engine`. should have been a separate commit in PR #76 to update the `vendor`

File: components/cli/vendor/github.com/docker/docker/opts/config.go (hunk 1 above)
Comment (andrewhsu) [inline components/cli/vendor/github.com/docker/docker/opts/config.go]: i traced the sequence of events and it looks like this missing `vendor` change is a result of creating the docker-ce 17.06 branch before the cli was re-vendored. here is the sequence of events:

- May 19: merge moby/moby#33296 to moby/moby which removed `opts/config.go`
- May 19: create docker-ce 17.06 branch
- Jul 7: re-vndr docker/docker to remove `opts/config.go` from `vendor` docker/cli@44ac808

File: components/cli/vendor/github.com/docker/docker/opts/opts.go (hunk 2 above)
Comment (andrewhsu) [inline components/cli/vendor/github.com/docker/docker/opts/opts.go]: same reason as https://github.com/docker/docker-ce/pull/116#discussion_r130779015

File: components/cli/vendor/github.com/docker/docker/opts/throttledevice.go (hunk 3 above)
Comment (andrewhsu) [inline components/cli/vendor/github.com/docker/docker/opts/throttledevice.go]: same reason as https://github.com/docker/docker-ce/pull/116#discussion_r130779015

File: components/cli/vendor/github.com/docker/docker/pkg/gitutils/gitutils.go (hunk 5 above)
Comment (andrewhsu) [inline components/cli/vendor/github.com/docker/docker/pkg/gitutils/gitutils.go]: cause of this discrepancy is same reason as https://github.com/docker/docker-ce/pull/116#discussion_r130778764: PR #100 should have had another commit to update `components/cli/vendor`

File: components/cli/vendor/github.com/docker/docker/pkg/term/tc_solaris_cgo.go (hunk 6 above)
Comment (andrewhsu) [inline components/cli/vendor/github.com/docker/docker/pkg/term/tc_solaris_cgo.go]: same reason as https://github.com/docker/docker-ce/pull/116#discussion_r130779015

File: components/cli/vendor/github.com/docker/docker/registry/config_unix.go (hunk 7 above)
Comment (andrewhsu) [inline components/cli/vendor/github.com/docker/docker/registry/config_unix.go]: cause of this discrepancy is same reason as https://github.com/docker/docker-ce/pull/116#discussion_r130778764: PR #71 should have had another commit to update `components/cli/vendor`

File: components/cli/vendor/github.com/docker/docker/volume/volume.go (hunk 8 above)
Comment (andrewhsu) [inline components/cli/vendor/github.com/docker/docker/volume/volume.go]: Looks like PR #48 should have had another commit to update `components/cli/vendor`

**Comment by andrewhsu** [comment]
I've scrubbed all of the comments put in by @thaJeztah (thanks for digging into it) and found that the discrepancies between `components/engine` and `components/cli/vendor/github.com/docker/docker` are caused by one of two reasons:

1. creating branch `17.06` when the upstream components docker/cli had not vendored moby/moby to keep in sync
2. creating PRs to `17.06` that modify code in `components/engine` but with no corresponding sync to vendor in `components/cli/vendor/github.com/docker/docker`

For the first item, I suggest we setup a regular vendor sync in docker/cli repo to vendor changes from moby/moby /cc @dnephin 

For the second item, I suggest we setup automatic vendor check to make sure code changes to `components/engine` are vendored into `components/cli/vendor/github.com/docker/docker` on PRs to docker-ce /cc @seemethere and I'd say #148 (thanks @tiborvass) is a step towards that direction.

So, I'm gong to close this PR in favor of #148 since it synced up the vendor directory a little later in time than this PR and caught 2 more vendor mismatches.

**Comment by dnephin** [comment]
Why is vendor mismatch a problem if the tests pass?

**Pull request closed without merging**